### PR TITLE
Dont assume any particular platforms

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
The MS Teams plugin currently isn't building windows or darwin. Instead of extending the hard-coded expectations for another plugin type, refactor to allow the plugin's own manifest to define what exectutables to expect and split apart.

Lots of discrete commits here to hopefully make things easier to review! Apologies that I had to upgrade to the public submodule, as I needed a version of `model.Manifest` that defined `Executables` as a map vs. the previous hard-coded set of just three platforms. A go version and slight tooling update later, and everything tests neatly.

#### Ticket Link
None.

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -19,6 +19,9 @@ linters-settings:
       - fieldalignment
   misspell:
     locale: US
+  staticcheck:
+    checks: ["all", "-SA1019"]
```
Comment (lieut-data) [inline .golangci.yml:23]: I fixed everything I could, then disabled while we still depend on `openpgp` here -- see https://github.com/golang/go/issues/44226.

File: server/jenkins.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ func getJenkins(jenkinsUser, jenkinsToken, jenkinsURL string) (*gojenkins.Jenkin
 	jenkins := gojenkins.CreateJenkins(jenkinsURL, jenkinsUser, jenkinsToken)
 	jenkins.Requester.SslVerify = config.SSLVerify
 	if config.CACrtPath {
-		caCert, err := ioutil.ReadFile("/app/ssl/ca.crt")
+		caCert, err := os.ReadFile("/app/ssl/ca.crt")
```
Comment (lieut-data) [inline server/jenkins.go:26]: Deprecated functionality.

File: server/plugin_release.go [hunk 3]
```diff
@@ -401,26 +401,13 @@ func verifySignatures(pluginFilePaths []string) error {
...
-	// If a plugin has special needs (e.g., it doesn't build for specific platforms), add that here:
-	if repositoryName == "mattermost-plugin-calls" {
-		expectedPlatformBinaries = map[string]string{
-			"linux-amd64":   "plugin-linux-amd64",
-			"freebsd-amd64": "plugin-freebsd-amd64",
-		}
-	}
-
-	if err := hasExpectedPlatformBinaries(pluginFilePath, expectedPlatformBinaries); err != nil {
+	platformBinaries, err := findPlatformBinaries(pluginFilePath)
```
Comment (lieut-data) [inline server/plugin_release.go:404]: Meat of changes start here.

File: server/plugin_release.go [hunk 4]
```diff
@@ -750,53 +751,57 @@ func getPluginSigningSftpClient(cfg *MatterbuildConfig) (*sftp.Client, error) {
...
+	err = unpackPlugin(filePath, tmpDir)
 	if err != nil {
-		return errors.Wrap(err, "failed to read gzip data")
+		return nil, errors.Wrap(err, "failed to unpack plugin")
 	}
 
-	foundBinaries := make(map[string]bool, len(expectedBinaries))
-	for _, v := range expectedBinaries {
-		foundBinaries[v] = false
+	dir, err := os.ReadDir(tmpDir)
```
Comment (lieut-data) [inline server/plugin_release.go:767]: This "pre-invocation setup" of `model.FindManifest` is copied right out of the server. It really should be part of `model.FindManifest` itself.

File: server/plugin_release_test.go [hunk 5]
```diff
@@ -366,24 +378,6 @@ func TestCreateTag(t *testing.T) {
 
 		gitMock.EXPECT().GetRefs(gomock.Eq(ctx), gomock.Eq(owner), gomock.Eq(repoName), gomock.Eq(fmt.Sprintf("tags/%s", tag))).Return(refs, nil, nil)
 
-		githubObj := &github.GitObject{
```
Comment (lieut-data) [inline server/plugin_release_test.go:369]: These assertions were invalid as the tag was never expected to be created. I believe the dependencies updates for the underlying mocking library "fixed" this.

File: server/server.go [hunk 6]
```diff
@@ -75,7 +75,7 @@ func Info(info string) {
 func WriteErrorResponse(w http.ResponseWriter, err *AppError) {
 	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(http.StatusOK)
-	w.Write([]byte(GenerateStandardSlashResponse(err.Error(), model.COMMAND_RESPONSE_TYPE_EPHEMERAL)))
+	w.Write([]byte(GenerateStandardSlashResponse(err.Error(), model.CommandResponseTypeEphemeral)))
```
Comment (lieut-data) [inline server/server.go:78]: v5 -> public submodule stuff.

File: server/test/mattermost-plugin-calls-v0.5.0.tar.gz
Comment (lieut-data) [inline server/test/mattermost-plugin-calls-v0.5.0.tar.gz:1]: Updating plugin.json to reflect the real binaries therein.

File: server/test/mattermost-plugin-demo-v0.4.1-linux-amd64.tar.gz
Comment (lieut-data) [inline server/test/mattermost-plugin-demo-v0.4.1-linux-amd64.tar.gz:1]: Updating plugin.json to reflect the real binaries therein.

**Comment by lieut-data** [comment]
@toninis & @mvitale1989, this is "mostly" developer code being reviewed, but I'm unsure of the permissions / ownership on this repo to know if that review is enough. Feel free to update reviewers as you see fit :)

**Review by cpoile — COMMENTED** [review]
Ah, I didn't realize you meant to remove Call's special case -- I think Calls still needs to have that here, unfortunately. We support building for six platforms (customer requests), but only officially support and release two of those.

**Comment by lieut-data** [comment]
> Ah, I didn't realize you meant to remove Call's special case -- I think Calls still needs to have that here, unfortunately. We support building for six platforms (customer requests), but only officially support and release two of those.

Hmm, the `plugin.json` for indeed defines six platforms:

https://github.com/mattermost/mattermost-plugin-calls/blob/fbea6f639f38d624752c0d655c55f61f1310d295/plugin.json#L9-L15

but then the binary actually ships four of these:

```sh
$ tar -tvzf com.mattermost.calls-0.22.1.tar.gz | grep 'server/dist/'
drwxr-xr-x  0 runner docker       0 12 Dec 15:00 com.mattermost.calls/server/dist/
-rwxr-xr-x  0 runner docker 31356821 12 Dec 15:00 com.mattermost.calls/server/dist/plugin-freebsd-amd64
-rwxr-xr-x  0 runner docker 31488874 12 Dec 15:00 com.mattermost.calls/server/dist/plugin-linux-amd64
-rwxr-xr-x  0 runner docker 31373654 12 Dec 15:00 com.mattermost.calls/server/dist/plugin-openbsd-amd64
-rwxr-xr-x  0 runner docker 30379704 12 Dec 15:00 com.mattermost.calls/server/dist/plugin-linux-arm64
```

To clarify, @cpoile, we only support two of the four we ship at present? Thoughts on just building all 6 and shipping all 6 but then continuing to only support 2 as we do today? (No special case would then be needed.)

**Comment by cpoile** [comment]
> To clarify, @cpoile, we only support two of the four we ship at present? Thoughts on just building all 6 and shipping all 6 but then continuing to only support 2 as we do today? (No special case would then be needed.)

My mistake -- we do package those 4 builds in ci, and publish them on github and through the marketplace. So we support those (2 are special cases for specific customers).

Let's try this PR then and we'll see how it goes. I'll have to keep a patch for development on darwin, but then with this PR matterbuild would be simpler.

As for building all 6 and shipping all 6, that's a no go -- we shouldn't ship plugins we don't support. :)

**Review by cpoile — APPROVED** [review]
Let's goooo

File: server/plugin_release.go [hunk 7]
```diff
@@ -443,33 +430,47 @@ func createPlatformPlugins(repositoryName, tag, pluginFilePath, pluginFolder str
...
 		return errors.Wrapf(err, "failed to decompress and extract to temporary directory")
 	}
 
+	return nil
+}
+
+// createPlatformPlugin takes a given plugin.tar.gz and creates a new one at the given path
+// for the given binary. Most plugin compilation steps generate a "omniplatform" bundle for
+// easy installation, but we strip out the unnecessary packages when pre-packaging
+// for a specific platform build of Mattermost.
```
Comment (streamer45) [inline server/plugin_release.go:461]: @cpoile I think this is the answer to your question.

File: server/plugin_release.go [hunk 8]
```diff
@@ -750,53 +751,57 @@ func getPluginSigningSftpClient(cfg *MatterbuildConfig) (*sftp.Client, error) {
...
-			break
-		}
-		if err != nil {
-			return errors.Wrap(err, "failed to read tar data")
-		}
+	// If the root of the plugin bundle consists of exactly one directory, assume the plugin
+	// is contained therein. Otherwise the root directory is expected to contain the plugin.
+	if len(dir) == 1 && dir[0].IsDir() {
+		tmpDir = filepath.Join(tmpDir, dir[0].Name())
+	}
```
Comment (streamer45) [inline server/plugin_release.go:776]: Mentioning this purely as an anecdote: this little check caused a bit of a headache to Chris a while back when we realized his local plugin bundles on MacOS included some OS specific hidden files that made this check fail mysteriously when installing the plugin.

File: server/plugin_release.go [hunk 9]
```diff
@@ -750,53 +751,57 @@ func getPluginSigningSftpClient(cfg *MatterbuildConfig) (*sftp.Client, error) {
...
+	if manifest.Server == nil {
+		return nil, fmt.Errorf("no server defined")
 	}
 
-	var missingBinaries []string
-	for k, v := range foundBinaries {
-		if !v {
-			missingBinaries = append(missingBinaries, k)
+	if len(manifest.Server.Executables) == 0 {
+		if len(manifest.Server.Executable) > 0 {
```
Comment (streamer45) [inline server/plugin_release.go:790]: Backwards compatibility? :exploding_head:

**Review by streamer45 — APPROVED** [review]
Looks good, thank you for the thorough comments throughout.

File: server/plugin_release.go (hunk 9 above)
Comment (lieut-data) [reply server/plugin_release.go:790]: Some developer, in a universe far, far away, will thank me for the clarity of this error message ;P

File: server/plugin_release.go (hunk 8 above)
Comment (lieut-data) [reply server/plugin_release.go:776]: Indeed, we should probably port https://github.com/mattermost/mattermost-plugin-playbooks/pull/1745 back to the starter template and push it out universally.

**Pull request merged**